File: 251002/Korti/publisher/src/providers/Sticker/Sticker.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseIntPipe,
  Post,
  Put,
} from '@nestjs/common';
import { StickerService } from './Sticker.service';
import { StickerRequestTo, UpdateStickerTo } from './Dto/StickerRequestTo';

@Controller('api/v1.0/stickers')
export class StickerController {
  constructor(private readonly stickerService: StickerService) {}

  @Get()
  @HttpCode(HttpStatus.OK)
  async getAll() {
    return await this.stickerService.getAllStickers();
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  async create(@Body() body: StickerRequestTo) {
    return await this.stickerService.createSticker(body);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param('id', ParseIntPipe) stickerId: number) {
    await this.stickerService.deleteSticker(stickerId);
  }

  @Get(':id')
  @HttpCode(HttpStatus.OK)
  async getOne(@Param('id', ParseIntPipe) stickerId: number) {
    return await this.stickerService.getStickerById(stickerId);
  }

  @Put()
  @HttpCode(HttpStatus.OK)
  async update(@Body() req: UpdateStickerTo) {
    return await this.stickerService.updateSticker(req);
  }

  @Get('articles/:stickerId')
  @HttpCode(HttpStatus.OK)
  async getByArticles(@Param('id', ParseIntPipe) id: number) {
    return await this.stickerService.getStickersByArticleId(id);
  }
}

File: 251003/Stasevichy/publisher/src/main/java/com/lab/labDCP/controller/StickerController.java
package com.lab.labDCP.controller;

import com.lab.labDCP.dto.*;
import com.lab.labDCP.entity.Sticker;
import com.lab.labDCP.service.StickerService;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Collections;
import java.util.List;
import java.util.NoSuchElementException;

@RestController
@RequestMapping("/api/v1.0/stickers")
public class StickerController {

    private final StickerService stickerService;

    public StickerController(StickerService stickerService) {
        this.stickerService = stickerService;
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public ResponseEntity<StickerResponseTo> createSticker(@RequestBody StickerRequestTo stickerRequestTo) {
        try {
            if(!validateSticker(stickerRequestTo)){
                return ResponseEntity.status(HttpStatus.FORBIDDEN).body(new StickerResponseTo());
            }
            Long id = System.currentTimeMillis();
            Sticker sticker = new Sticker(id, stickerRequestTo.getName());
            StickerResponseTo response = stickerService.createSticker(sticker);
            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        }catch (DataIntegrityViolationException e){
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body(new StickerResponseTo());
        }
        catch (Exception e){
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(new StickerResponseTo());
        }
    }

    @GetMapping
    public ResponseEntity<List<StickerResponseTo>> getAllStickers() {
        try {
            List<StickerResponseTo> response = stickerService.getAllStickers();
            return ResponseEntity.status(HttpStatus.OK).body(response);
        }catch (Exception e){
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Collections.emptyList());
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<StickerResponseTo> getStickerById(@PathVariable Long id) {
        try {
            StickerResponseTo response = stickerService.getStickerById(id);
            return ResponseEntity.status(HttpStatus.OK).body(response);
        }catch (Exception e){
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(new StickerResponseTo());
        }
    }

    @PutMapping
    public ResponseEntity<StickerResponseTo> updateNews(@RequestBody StickerRequestTo stickerRequestTo) {
        try {
            if (stickerRequestTo.getId() == null) return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new StickerResponseTo());

            StickerResponseTo updatedSticker = stickerService.updateSticker(stickerRequestTo.getId(), stickerRequestTo);
            return ResponseEntity.ok(updatedSticker);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new StickerResponseTo());
        }
        catch (IllegalArgumentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new StickerResponseTo());
        }
        catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(new StickerResponseTo());
        }
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public ResponseEntity<Void> deleteSticker(@PathVariable Long id) {
        try {
            stickerService.deleteSticker(id);
            return ResponseEntity.noContent().build();
        } catch (NoSuchElementException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        } catch (Exception e){
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }
    private boolean validateSticker(StickerRequestTo stickerRequestTo) {

        if (stickerRequestTo.getName().length() < 2 ||
                stickerRequestTo.getName().length() > 32) {
            return false;
        }

        return true;
    }
}

